lms/static/scripts/frontend_apps/components/BasicLtiLaunchApp.js

```diff
@@ -23,105 +25,155 @@ import Spinner from './Spinner';
+      // to relay to the sidebar via RPC.
+      sync: apiSync,
     },
     grading,
     // Content URL to show in the iframe.
     viaUrl,
     canvas,
   } = useContext(Config);
 
-  const [ltiLaunchState, setLtiLaunchState] = useState({
-    ...INITIAL_LTI_LAUNCH_STATE,
-    state: viaCallbackUrl ? 'fetching-url' : 'fetched-url',
-    contentUrl: viaUrl ? viaUrl : null,
-  });
+  // The current state of the error.
+  // One "error-fetch", "error-authorizing", or "error-report-submission"
+  const [errorState, setErrorState] = useState(null);
+
+  // Any current error event object thrown.
+  const [error, setError] = useState(null);
+
+  // When the app is initially displayed, it will use the Via URL if given
+  // or invoke the API callback to fetch the URL otherwise.
+  const [contentUrl, setContentUrl] = useState(viaUrl ? viaUrl : null);
+
+  // How many current blocking request are pending
+  const [fetchCount, setFetchCount] = useState(0);
+
+  // Var to keep track of the fetchCount
+  const count = useRef(0);
+
+  // `AuthWindow` instance, set only when waiting for the user to approve
+  // the app's access to the user's files in the LMS.
+  const authWindow = useRef(null);
+
+  // Show the assignment when there are no pending requests
+  // and errorState is falsely
+  const showIframe = () => {
+    return fetchCount === 0 && !errorState;
+  };
+
+  // Show the loader is there are any pending requests and
+  // and errorState is falsely
+  const showSpinner = () => {
+    return !errorState && fetchCount > 0;
+  };
+
+  // Increment the fetch counter by 1 and clear any
+  // previous error state.
+  const setFetching = () => {
+    count.current += 1;
+    setFetchCount(count.current);
+    setErrorState(null);
+  };
+
+  // Decrement the fetch counter by 1.
+  const setFetched = () => {
+    count.current -= 1;
+    setFetchCount(count.current);
+  };
+
+  /**
+   * Helper to handle thrown errors from from API requests.
+   *
+   * @param {Error} e - Error object from request.
+   * @param {string} errorState - One of "error-fetch",
+   *  "error-authorizing", or "error-report-submission"
+   * @param {boolean} [retry=true] - Can the request be retried?
+   */
+  const handleError = (e, errorState, retry = true) => {
+    setFetched();
+    if (e instanceof ApiError && !e.errorMessage && retry) {
+      setErrorState('error-authorizing');
+    } else {
+      setError(e);
+      setErrorState(errorState);
+    }
+  };
+
+  /**
+   * Fetch the groups from the sync endpoint if `sync` object exists.
+   */
+  const fetchGroups = async () => {
+    if (apiSync) {
+      try {
+        setFetching();
+        const groups = await apiCall({
+          authToken,
+          path: apiSync.path,
+          data: apiSync.data,
+        });
+        rpcServer.resolveGroupFetch(groups);
+        setFetched();
+      } catch (e) {
+        handleError(e, 'error-fetch');
```

Comment:
If an exception is thrown, this will leave the `fetchCount` state with a value > 0. One way to resolve that would be decrementing the counter in a `finally` block.